rust/src/adapter/zipkin.rs

```diff
@@ -0,0 +1,141 @@
+
+        Ok(ZipkinTraceCtx(collector))
+    }
+}
+
+#[derive(Clone)]
+pub struct ZipkinAdapter {
+    pub trace_id: String,
+    pub spans: Vec<Span>,
+}
+
+impl ZipkinAdapter {
+    #[allow(clippy::new_ret_no_self)]
+    pub fn new() -> ZipkinAdapterContainer {
+        let adapter = Self {
+            trace_id: new_span_id().to_hex_8(),
+            spans: vec![],
+        };
+
+        ZipkinAdapterContainer(Arc::new(Mutex::new(adapter)))
+    }
+
+    fn _handle_event(&mut self, event: Event, parent_span_id: Option<String>) -> Option<Vec<Span>> {
+        match event {
+            Event::Func(_id, f) => {
+                let name = f.name.clone().unwrap_or("unknown-name".to_string());
+
+                let span =
+                    Span::new(self.trace_id.clone(), parent_span_id, name, f.start, f.end);
+
+                let span_id = Some(span.id.clone());
+                let mut spans = vec![span];
+
+                for e in f.within.iter() {
+                    if let Some(mut s) = self._handle_event(e.to_owned(), span_id.clone()) {
+                        spans.append(&mut s);
+                    };
+                }
+
+                Some(spans)
+            }
+            Event::Alloc(_id, a) => {
+                let mut span = Span::new(
+                    self.trace_id.clone(),
+                    parent_span_id,
+                    "allocation".to_string(),
+                    a.ts,
+                    a.ts,
+                );
+                span.add_tag_i64("amount".to_string(), a.amount.into());
+                Some(vec![span])
+            }
+            Event::Metadata(_id, Metadata { key, value }) => {
+                if key == "trace_id" {
+                    self.trace_id = value.to_hex_8();
+                }
+
+                None
+            }
+            Event::Shutdown(_id) => {
+                if let Err(e) = self.shutdown() {
+                    warn!("Failed to shutdown Zipkin adapter {}", e);
+                }
+                self.spans.clear();
+                None
+            }
+        }
+    }
+}
+
+impl Adapter for ZipkinAdapter {
+    // flush any remaning spans
+    fn shutdown(&self) -> Result<()> {
+        let mut ztf = ZipkinFormatter::new();
+        ztf.spans = self.spans.clone();
+
+        let mut first_span = ztf.spans.first_mut().unwrap();
+        first_span.local_endpoint = Some(LocalEndpoint { service_name: Some("my_service".into()) });
+
+        let url = "http://localhost:9411/api/v2/spans";
+        let j = json!(&ztf.spans);
+        let body = serde_json::to_string(&j)?;
+
+        // perhaps this should be an async operation with something
+        // like reqwest?
+        let response = ureq::post(url)
+            .timeout(Duration::from_secs(1))
+            .set("Content-Type", "application/json")
+            .send_string(&body);
+
+        // should maybe retry or throw an exception
+        // TODO check default logic of http client
```

Comment:
This should return an error instead of Ok() on line 131. There isn't much the user can do with this knowledge other than alerting that Zipkin is unhealthy, but it's good to give them the option of knowing that.